### PR TITLE
Fix 0/0 issue

### LINKED ISSUE #484: 'nan' problem in d1 and BlackScholes.price
This hardly happens but I faced this issue:

```
import torch
from pfhedge.instruments import EuropeanOption
from pfhedge.instruments import BrownianStock
derivative = EuropeanOption(BrownianStock(), maturity=5/250)
from pfhedge.nn import BlackScholes

BlackScholes(derivative).price(log_moneyness=torch.tensor(1), time_to_maturity=torch.tensor(0), volatility=torch.tensor(0.2))
# works normally: tensor(1.7183)
BlackScholes(derivative).price(log_moneyness=torch.tensor(0), time_to_maturity=torch.tensor(0), volatility=torch.tensor(0.2))
# error

BlackScholes(derivative).price(log_moneyness=torch.tensor(1e-7), time_to_maturity=torch.tensor(0), volatility=torch.tensor(0.2))
# tensor(1.1921e-07)
BlackScholes(derivative).price(log_moneyness=torch.tensor(-1e-7), time_to_maturity=torch.tensor(0), volatility=torch.tensor(0.2))
# tensor(0.)
```
Considering the continuity of pricing, the error case should return tensor of 0.

This issue comes from the problem in black-sholes d1 calculation:

https://github.com/pfnet-research/pfhedge/blob/c342b52ba2d8c6182304a20a0db21e83130799ef/pfhedge/nn/modules/bs/european.py#L222
https://github.com/pfnet-research/pfhedge/blob/c342b52ba2d8c6182304a20a0db21e83130799ef/pfhedge/nn/functional.py#L521
-> `(0.0).div(0.0)` returns `nan`
Then, 
https://github.com/pfnet-research/pfhedge/blob/c342b52ba2d8c6182304a20a0db21e83130799ef/pfhedge/nn/functional.py#L473
https://github.com/pytorch/pytorch/blob/71f889c7d265b9636b93ede9d651c0a9c4bee191/torch/distributions/normal.py#L79-L82
in `self._validate_sample(value)`,
https://github.com/pytorch/pytorch/blob/71f889c7d265b9636b93ede9d651c0a9c4bee191/torch/distributions/distribution.py#L286-L294
these validations returns `ValueError` because of nan.

I think,
```
BlackScholes(derivative).price(log_moneyness=torch.tensor(0), time_to_maturity=torch.tensor(0), volatility=torch.tensor(0.2))
```
should return tensor of 0, but, I'm not sure if we should fix the calculation in d1 or make an exceptional routine for this edge case.

### PR DESCRIPTION
**Pull request by masanorihirano**
fix #484 

This fix includes:
 - full fix for BS d1/d2, European options with additional test
 - partial fix for European binary (because the current version mainly uses autogreeks) **w/o additional test**

This fix also fixes the lookback option indirectly.

Because other options' BS modules mainly use autogreeks, I didn't make modifications except for adding notes.

### CODE REVIEW

File: pfhedge/nn/functional.py [hunk 1]
```diff
@@ -518,7 +518,19 @@ def d1(log_moneyness: Tensor, time_to_maturity: Tensor, volatility: Tensor) -> T
...
+        raise ValueError("all elements in volatility have to be non-negative")
+    numerator = s + (v.square() / 2) * t
+    denominator = v * t.sqrt()
+    zero_div_zero_mask = (numerator == 0).logical_and(denominator == 0)
+    denominator = torch.where(
+        zero_div_zero_mask,
+        torch.zeros_like(denominator) + torch.finfo().tiny,
+        denominator,
+    )
+    return numerator.div(denominator)
```
Comment (simaki) [inline pfhedge/nn/functional.py:533]: Nice that nonzero/0 is inf rather than a large but finite number (expected behavior right?)

NIT: Possibly it could be a bit more intuitive to express it like:

output = numerator / denominator
output.where(numerator == 0 and denominator == 0, 0)

File: pfhedge/nn/functional.py [hunk 2]
```diff
@@ -518,7 +518,19 @@ def d1(log_moneyness: Tensor, time_to_maturity: Tensor, volatility: Tensor) -> T
         torch.Tensor
     """
     s, t, v = broadcast_all(log_moneyness, time_to_maturity, volatility)
-    return (s + (v.square() / 2) * t).div(v * t.sqrt())
+    if not (t >= 0).all():
+        raise ValueError("all elements in time_to_maturity have to be non-negative")
+    if not (v >= 0).all():
+        raise ValueError("all elements in volatility have to be non-negative")
```
Comment (simaki) [inline pfhedge/nn/functional.py:524]: Nice

**Review by simaki — APPROVED** [review]
Perfect!
I leave it to you whether to incorporate some nits.

Will merge in a week or so. Lots of thanks!

File: pfhedge/nn/functional.py (hunk 1 above)
Comment (masanorihirano) [reply pfhedge/nn/functional.py:533]: > Nice that nonzero/0 is inf rather than a large but finite number (expected behavior right?)

Yes, I meant to.

According to your suggestion, I'll fix

**Comment by simaki** [comment]
👍
